### PR TITLE
Navigation: Implement Keyboard Navigation

### LINKED ISSUE #38454: Navigation: Add keyboard navigation to menus and submenus
**What happened:** 

The first level navigation submenus are not accessible via the keyboard, there are some [guides](https://www.w3.org/WAI/tutorials/menus/flyout/) from WCAG to address this problem.

But unfortunately, this guide is misaligned with our current design:

- Our menu items are also a link that redirects to a different page
- We don't have a button that serves as a visual indicator for the presence of a submenu.

Alternatives to solve the problem:

- Improve our UX and add a toggle button for submenus -- Navigation is through tabs, we will use flyout menus
- Follow a different approach to make the menu accessible, something like treating our navigation menu as a [Web Application Menu](https://www.w3.org/WAI/tutorials/menus/application-menus-code/) -- Here navigation is through arrows keys, and not using tab tab

**What you expected to happen:** 

- [ ] User is able to navigate the first level menu using keyboard (Tab)
- [ ] User visually is aware there are submenus when navigating first level navigation (When tabbing the submenu fly out)
- [ ] User is able to navigate inside submenu using arrow keys, Arrow Right-> Focus first item submenu, Arrows up and down -> Navigate submenu
- [ ] User is not forced to move through all submenus to switch to another first-level menu, it can use Esc, Tab, or the left arrow key
- [ ] User should be able to move away from the menu to main section (Skip links)
- [ ] Make sure this work doesn't interfere with Enterprise
    - [ ] Check that it compiles with Enterprise
    - [ ] Check with someone on Enterprise that we haven't broken anything 

**How to reproduce it (as minimally and precisely as possible):**

### PR DESCRIPTION
**Pull request by axelavargas**
**What this PR does / why we need it**:

Implement keyboard navigation to NavBar menu and submenus, introduces react-aria in our current architecture

Fixes that are missing:

- [x] Styling: Fix offset on the focus ring for menu item elements ([Issue](https://github.com/grafana/grafana/issues/42750))
- [x] Styling: Fix focus ring visible on hover after navigating through keyboard  [Issue](https://github.com/grafana/grafana/issues/42751)
- [x] Behavior: Arrow left in subitem should focus on parent  (Future PR)
- [x] Behavior:  Divider should not be focusable
- [x] Behaviour: Arrow right should open submenu if menu is closed
- [x] Behaviour: Focus on reverse direction is not working properly 
- [x] Behaviour: Subtitle is focusable
- [x] Remove unused CSS
- [x] Add types related to react aria on NavBarItemButton 
- [x] Fix click when interacting to onHover
- [x] Add unit test 

Important
Make sure this work doesn't interfere with Enterprise
- [x] Check that it compiles with Enterprise
- [x] Check with someone on Enterprise that we haven't broken anything

**Which issue(s) this PR fixes**:

Fixes #38454

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by torkelo** [comment]
Great progress! 

I know it's draft but wanted to see it in action to took it for a spin, and discovered some thing you might already have noticed yourself: 

Can focus the divider: 
Screenshot from 2021-11-12 08-59-55

When clicking on the page it hides the menu but the focus of the navbar item stays put (even when clicking again and navigating to another page). First on second click is focus cleared. The same is true if I hit enter key on keyboard, I navigate to page but the navbar item still has focus

**Comment by axelavargas** [comment]
> Great progress!
> 
> I know it's draft but wanted to see it in action to took it for a spin, and discovered some thing you might already have noticed yourself:
> 
> Can focus the divider: Screenshot from 2021-11-12 08-59-55
> 
> When clicking on the page it hides the menu but the focus of the navbar item stays put (even when clicking again and navigating to another page). First on second click is focus cleared. The same is true if I hit enter key on keyboard, I navigate to page but the navbar item still has focus

Thank you, Torkel for giving it a spin 🙌🏾, I'll fix it

File: public/app/core/components/NavBar/NavBarItem.test.tsx [hunk 1]
```diff
@@ -1,55 +1,144 @@
...
+      });
+    });
+
+    describe('and hovering over the menu trigger button', () => {
+      it('then the menu items should be visible', () => {
+        getTestContext();
+
+        userEvent.hover(screen.getByRole('button'));
+
+        expect(screen.getByText('Parent Node')).toBeInTheDocument();
```
Comment (Clarity-89) [inline public/app/core/components/NavBar/NavBarItem.test.tsx:57]:
```suggestion
        expect(screen.getByRole('menuitem', { name: 'Parent Node' })).toBeInTheDocument();
```
I think we can test this and other menu items with more accessibility in mind, as they seem to have `menuitem` role.

**Review by Clarity-89 — COMMENTED** [review]
Tested this with Enterprise and didn't find any issues.

Left a few suggestions regarding tests, also it seems like some tests produce a warning: 
```
    console.error
      Warning: An update to NavBarItemMenuTrigger inside a test was not wrapped in act(...).
      
      When testing, code that causes React state updates should be wrapped into act(...):
      
      act(() => {
        /* fire events that update state */
      });
      /* assert on the output */
```

Maybe the tests involving `NavBarItemMenuTrigger` should be made async, which has proven to remove this warning in the past.

**Comment by axelavargas** [comment]
> Tested this with Enterprise and didn't find any issues.
> 
> Left a few suggestions regarding tests, also it seems like some tests produce a warning:
> 
> ```
>     console.error
>       Warning: An update to NavBarItemMenuTrigger inside a test was not wrapped in act(...).
>       
>       When testing, code that causes React state updates should be wrapped into act(...):
>       
>       act(() => {
>         /* fire events that update state */
>       });
>       /* assert on the output */
> ```
> 
> Maybe the tests involving `NavBarItemMenuTrigger` should be made async, which has proven to remove this warning in the past.

Hey @Clarity-89 👋🏾 , thank you for taking a look 🙌🏾 , I will check today at the test suggestions, as an FYI we are planning to tackle a couple of minor things in a future PR and merge this branch early into the main branch to give enough time for testing :) (We aim to release on 8.4)

File: public/app/core/components/NavBar/NavBarItemMenuTrigger.tsx [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+  let { isFocusVisible } = useFocusVisible({ isTextInput: false });
+
+  const { hoverProps } = useHover({
+    onHoverChange: (isHovering) => {
+      if (isHovering) {
+        state.open();
+      }
+      if (!isHovering) {
+        state.close();
+      }
```
Comment (kaydelaney) [inline public/app/core/components/NavBar/NavBarItemMenuTrigger.tsx:45]: Nit:
```suggestion
      } else {
        state.close();
      }
```

**Review by kaydelaney — APPROVED** [review]
Very nice work with this! Just left a tiny comment, but otherwise looks great. 👍

File: public/app/core/components/NavBar/utils.ts [hunk 3]
```diff
@@ -136,3 +136,7 @@ export const isSearchActive = (location: Location<unknown>) => {
   const query = new URLSearchParams(location.search);
   return query.get('search') === 'open';
 };
+
+export function getNavModelItemKey(item: NavModelItem) {
+  return `${item.id ?? item.text}`;
```
Comment (ashharrison90) [inline public/app/core/components/NavBar/utils.ts:141]:
```suggestion
  return item.id ?? item.text;
```

File: public/app/core/components/NavBar/NavBarMenuItem.tsx [hunk 4]
```diff
@@ -31,25 +43,36 @@ export function NavBarMenuItem({ icon, isActive, isDivider, onClick, styleOverri
...
+    ) : (
+      <li>{element}</li>
+    );
+  }
 
-  return isDivider ? <li data-testid="dropdown-child-divider" className={styles.divider} /> : <li>{element}</li>;
+  return isDivider ? (
+    <div data-testid="dropdown-child-divider" className={styles.divider} tabIndex={-1} aria-disabled />
+  ) : (
+    <React.Fragment>{element}</React.Fragment>
```
Comment (ashharrison90) [inline public/app/core/components/NavBar/NavBarMenuItem.tsx:74]: minor nit
```suggestion
    <>{element}</>
```

**Review by ashharrison90 — APPROVED** [review]
nice! 🙌  couple of minor nits, and i'm also seeing the focus styles on hover sometimes

image

**Comment by axelavargas** [comment]
> nice! raised_hands couple of minor nits, and i'm also seeing the focus styles on hover sometimes
> 
> image

Thank you Ash for taking a look 🙌🏾 , the on hover issue will be tackled in another PR :)

**Review by nevermind89x — APPROVED** [review]
Reviewed for the Onboarding plugin, this looks good to me, and nothing that would break there. Thanks!

**Pull request merged**